src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+            }
+
+            if (size == totalSize && size == 1)
+            {
+                return true;
+            }
+
+            return size <= totalSize * _anomalyRatioThreshold;
+        }
+
+        private double GetDimensionEntropy(Dictionary<string, int> pointDis, Dictionary<string, int> anomalyDis)
+        {
+            int total = GetTotalNumber(pointDis);
+            double entropy = 0;
+
+            foreach (string key in anomalyDis.Keys)
+            {
+                double dimEntropy = GetEntropy(pointDis[key], anomalyDis[key]);
+                entropy += dimEntropy * pointDis[key] / total;
+            }
+
+            return entropy;
+        }
+
+        private double GetDimensionInstrinsicValue(Dictionary<string, int> pointDis)
+        {
+            int total = GetTotalNumber(pointDis);
+            double instrinsicValue = 0;
+
+            foreach (string key in pointDis.Keys)
+            {
+                instrinsicValue -= Log2((double)pointDis[key] / total) * (double)pointDis[key] / total;
+            }
+
+            return instrinsicValue;
+        }
+
+        private int GetTotalNumber(Dictionary<string, int> distribution)
+        {
+            int total = 0;
+            foreach (int num in distribution.Values)
+            {
+                total += num;
+            }
+            return total;
+        }
+
+        private void UpdateDistribution(Dictionary<string, int> distribution, List<Point> points, string dimKey)
+        {
+            foreach (Point point in points)
+            {
+                string dimVal = (string)point.Dimension[dimKey];
+                if (!distribution.ContainsKey(dimVal))
+                {
+                    distribution.Add(dimVal, 0);
+                }
+                distribution[dimVal] = distribution[dimVal] + 1;
+            }
+        }
+
+        private double Log2(double val)
+        {
+            if (Double.IsNaN(val))
+            {
+                return 0;
+            }
+
+            return Math.Log(val) / Math.Log(2);
+        }
+
+        private static bool ContainsAll(Dictionary<string, Object> bigDic, Dictionary<string, Object> smallDic)
+        {
+            foreach (var item in smallDic)
+            {
+                if (!bigDic.ContainsKey(item.Key))
+                {
+                    return false;
+                }
+
+                if (bigDic.ContainsKey(item.Key) && !bigDic[item.Key].Equals(smallDic[item.Key]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool IsAggregationDimension(Object val, Object aggSymbol)
+        {
+            return val.Equals(aggSymbol);
+        }
+    }
+
+    public class DimensionInfo
+    {
+        public List<string> DetailDims { get; set; }
+        public List<string> AggDims { get; set; }
+
+        public static DimensionInfo CreateDefaultInstance()
+        {
```

Comment:
question: Why not use a regular constructor?

Reply:
> question: Why not use a regular constructor?

updated

Reply:
Sorry, I still see a static function.  Where is this updated?

---
In reply to: [422135070](https://github.com/dotnet/machinelearning/pull/4925#discussion_r422135070) [](ancestors = 422135070)

Reply:
Can you please update this?

---
In reply to: [422244730](https://github.com/dotnet/machinelearning/pull/4925#discussion_r422244730) [](ancestors = 422244730,422135070)